rules/TypeDeclaration/Rector/Empty_/EmptyOnNullableObjectToInstanceOfRector.php

```diff
@@ -86,7 +86,7 @@ public function refactor(Node $node): ?Node
             return null;
         }
 
-        $exprType = $this->getType($empty->expr);
+        $exprType = $this->nodeTypeResolver->getNativeType($empty->expr);
```

Comment:
maybe we should have a new `getNativeType()` on `AbstractRector`?

Reply:
I will let @TomasVotruba decide on that, as the goal should be less method in `AbstractRector`.

Reply:
yeah I agree (its not important for this PR, as it would be a followup).

I think we should either have both on AbstractRector or none.
otherwise people will use `getType()` and don't think about phpdoc types

Reply:
for PHPStan, `getType()` is got correct type on native function, eg:

```php
        $parts = parse_url($url);

        if (!empty($parts['host'])) {
            return $parts['host'];
        }

        return null;
```

the `$parts['host']` goes correct type: `StringType`, while when using:

```php
$this->nodeTypeResolver->getNativeType()
```

that goes `MixedType`.

see 

https://github.com/rectorphp/rector-src/blob/8ba3d90ba430782c8e9bba135411260df802df3e/rules/Strict/Rector/Empty_/DisallowedEmptyRuleFixerRector.php#L126-L128

on fixture:

https://github.com/rectorphp/rector-src/blob/8ba3d90ba430782c8e9bba135411260df802df3e/tests/Issues/EmptyBooleanCompare/Fixture/fixture.php.inc#L5-L37

Also, `getType()` is usefull when:

1. we don't want to change the code if already using docblock definition
2. on downgrade, that type downgrade follow up to docblock should stay have the same type so allowed to be downgraded correctly on continuation by other rules.

Reply:
Also, this change seems `$empty->expr` is checking `ArrayDimFetch` on this case, see above check on `ArrayDimFetch`, so the new fixture is unrelated with the change :)

`ArrayDimFetch` is mostly correct as defining array key type defined in docblock

Reply:
I don't say `getType()` should never be used, but promoting one while don't the other means people will not realize they need to think about it

Reply:
> Also, this change seems `$empty->expr` is checking `ArrayDimFetch` on this case, see above check on `ArrayDimFetch`, so the new fixture is unrelated with the change :)
> 
> `ArrayDimFetch` is mostly correct as defining array key type defined in docblock

if you revert my change, the added test fails

Reply:
ah, I see, that truthy check before :), I removed my comment, you already see it

Reply:
Yeah, probably worth a blog post :)